buildkite/bazel-bench/bazel_bench.py

```diff
@@ -276,35 +332,43 @@ def main(args=None):
     for project in PROJECTS:
         if not project["active"]:
             continue
-        platforms = _get_platforms(project["name"])
+        platforms = _get_platforms(
+            project["name"], whitelist=PLATFORMS_WHITELIST)
         for platform in platforms:
-            # bazel-bench doesn't support Windows for now.
-            if platform in ["windows"]:
-                continue
-
             # When running on the first platform, get the bazel commits.
             # The bazel commits should be the same regardless of platform.
             if not bazel_commits:
                 bazel_clone_path = bazelci.clone_git_repository(BAZEL_REPOSITORY, platform)
-                bazel_commits = _get_bazel_commits(date, bazel_clone_path)
+                bazel_commits = _get_bazel_commits(
+                    date, bazel_clone_path, parsed_args.max_commits)
 
             bazel_bench_ci_steps.append(
                 _ci_step_for_platform_and_commits(
-                    bazel_commits, platform, project, parsed_args.bazel_bench_options, date
+                    bazel_commits, platform, project,
+                    parsed_args.bazel_bench_options, date, parsed_args.bucket
                 )
             )
         _create_and_upload_metadata(
             project_label=project["storage_subdir"],
+            project_source=project["git_repository"],
             command=project["bazel_command"],
             date=date,
             platforms=platforms,
+            bucket=parsed_args.bucket
         )
 
-    bazelci.eprint(yaml.dump({"steps": bazel_bench_ci_steps}))
-    buildkite_pipeline_cmd = "cat <<EOF | buildkite-agent pipeline upload\n%s\nEOF" % yaml.dump(
-        {"steps": bazel_bench_ci_steps}
-    )
-    subprocess.call(buildkite_pipeline_cmd, shell=True)
+        bazel_bench_ci_steps.append("wait")
+        # If all the above steps succeed, generate the report.
+        bazel_bench_ci_steps.append(
+            _report_generation_step(
+                date, project["storage_subdir"],
+                parsed_args.bucket, REPORT_GENERATION_PLATFORM))
+
+        bazelci.eprint(yaml.dump({"steps": bazel_bench_ci_steps}))
+        buildkite_pipeline_cmd = "cat <<EOF | buildkite-agent pipeline upload\n%s\nEOF" % yaml.dump(
+            {"steps": bazel_bench_ci_steps}
+        )
+        subprocess.call(buildkite_pipeline_cmd, shell=True)
```

Comment:
You can replace this shell-injection prone way of doing it by just a single call where you pipe in the data directly:

```python
subprocess.run(["buildkite-agent", "pipeline", "upload"], input=yaml.dump({"steps": bazel_bench_ci_steps}))
```

Much cleaner and safer and doesn't require a shell. :)

Reply:
Done. TIL.